File: src/plugLog/libraries/LogComponentBaseJava/LogComponentBase.java
/****************************************************************************/
/* LogComponentBase class                                                   */
/*                                                                          */
/*  Author(s):                                                              */
/*    - Georg Hoesch (hoesch@in.tum.de)                                     */
/*    - Cyrille Pontvieux (cyrille.pontvieux@edu.univ-fcomte.fr)            */
/*                                                                          */
/* $LICENSE$                                                                */
/****************************************************************************/
/* $Id: LogComponentBase.java,v 1.1 2004/01/09 11:07:12 ghoesch Exp $
 * $Log: LogComponentBase.java,v $
 * Revision 1.1  2004/01/09 11:07:12  ghoesch
 * Restructured the whole LogService source tree.
 * Added autotools make process. Cleaned up code.
 * Removed some testers. Ready to release.
 *
 * Revision 1.6  2003/12/12 19:36:13  cpontvie
 * New version of ORBTools (using the POA bidir in C++) and fully handle the
 * servants activation.
 *
 * Revision 1.5  2003/11/21 11:02:27  cpontvie
 * Catch more exceptions, that allow the component/tool to not crash if the
 * LogCentral crash.
 *
 * Revision 1.4  2003/11/20 09:15:54  cpontvie
 * Remove a debug message
 *
 * Revision 1.3  2003/11/19 17:51:44  cpontvie
 * New version of the LogComponentBase in Java that correct some bugs in the
 * initalization and in the sendMsg method.
 *
 * Revision 1.2  2003/11/18 18:41:14  cpontvie
 * Catch more exceptions
 *
 * Revision 1.1  2003/11/18 10:51:32  cpontvie
 * Add the LogComponentBase libraries
 *
 ****************************************************************************/

import org.omg.CosNaming.*;
import org.omg.CosNaming.NamingContextPackage.*;
import org.omg.CORBA.*;
import org.omg.PortableServer.*;
import org.omg.PortableServer.POA;
import java.util.*;
import java.net.InetAddress;

public class LogComponentBase
{
  public
  LogComponentBase(boolean_out success, String[] args,
                   int tracelevel, int port) {
    success.value = false;
    this.partialInit = false;
    this.name = "";
    try {
      InetAddress ia[] = InetAddress.getAllByName(
        InetAddress.getLocalHost().getHostName()
      );
      if (ia.length != 0) {
        this.hostname = ia[0].getHostName();
      } else {
        this.hostname = "unknownhost";
      }
    } catch (java.net.UnknownHostException ex) {
      this.hostname = "unknownhost";
    }
    if (!ORBTools.init(args, tracelevel, port)) {
      System.out.println("Cannot initialize the ORB !");
      return;
    }
    CORBA_Object_out objout = new CORBA_Object_out();
    if (!ORBTools.findServant("LogService", "", "LogComponent", "", objout)) {
      System.out.println("Cannot contact the LogComponent servant of "
        + "LogService (not found) !");
      return;
    }
    this.LCCref = LogCentralComponentHelper.narrow(objout.value);
    this.CCimpl = new ComponentConfiguratorImpl(this);
    if (!ORBTools.activateServant(this.CCimpl, this.id)) {
      System.out.println("Cannot activate the ComponentConfigurator servant !");
      return;
    }
    if (!ORBTools.activatePOA()) {
      System.out.println("Cannot activate the POA !");
      return;
    }
    this.CCref = this.CCimpl._this();
    this.pingThread = null;
    this.flushBufferThread = null;
    try { // To make sure that the object is callable
      this.LCCref.test();
    } catch (Exception ex2) {
      System.out.println("Cannot contact the LogComponent servant of "
        + "LogService, even if it was found !");
      return;
    }
    this.currentTagList = new String[0];
    this.buffer = new log_msg_t[0];
    success.value = true;
  }

  public
  LogComponentBase(boolean_out success, org.omg.CORBA.ORB orb) {
    success.value = false;
    this.partialInit = true;
    this.name = "";
    try {
      InetAddress ia[] = InetAddress.getAllByName(
        InetAddress.getLocalHost().getHostName()
      );
      if (ia.length != 0) {
        this.hostname = ia[0].getHostName();
      } else {
        this.hostname = "unknownhost";
      }
    } catch (java.net.UnknownHostException ex) {
      this.hostname = "unknownhost";
    }
    ORBTools.orb = orb;

    //find the POA
    try {
      org.omg.CORBA.Object objRef =
        orb.resolve_initial_references("RootPOA");
      try {
        ORBTools.poa = org.omg.PortableServer.POAHelper.narrow(objRef);
        ORBTools.pman = ORBTools.poa.the_POAManager();
      } catch(org.omg.CORBA.SystemException ex) {
        System.out.println("Failed to find the POA.");
        return;
      }
    } catch(org.omg.CORBA.ORBPackage.InvalidName ex1) {
      return;
    }
    // Activate the POA if it is not already activate
    ORBTools.activatePOA();

    CORBA_Object_out objout = new CORBA_Object_out();
    if (!ORBTools.findServant("LogService", "", "LogComponent", "", objout)) {
      System.out.println("Cannot contact the LogComponent servant of "
        + "LogService (not found) !");
      return;
    }
    this.LCCref = LogCentralComponentHelper.narrow(objout.value);
    this.CCimpl = new ComponentConfiguratorImpl(this);
    if (!ORBTools.activateServant(this.CCimpl, this.id)) {
      System.out.println("Cannot activate the ComponentConfigurator servant !");
      return;
    }
    this.CCref = this.CCimpl._this();
    this.pingThread = null;
    this.flushBufferThread = null;
    try { // To make sure that the object is callable
      this.LCCref.test();
    } catch (Exception ex2) {
      System.out.println("Cannot contact the LogComponent servant of "
        + "LogService, even if it was found !");
      return;
    }
    this.currentTagList = new String[0];
    this.buffer = new log_msg_t[0];
    success.value = true;
  }

  // kind of destructor
  public
  void
  finalize() {
    if (this.pingThread != null) {
      this.pingThread.stopThread();
      this.pingThread = null;
    }
    if (this.flushBufferThread != null) {
      this.flushBufferThread.stopThread();
      this.flushBufferThread = null;
    }
    this.name = null;
    this.hostname = null;
    this.CCimpl = null;
    ORBTools.deactivateServant(this.id);
    if (!this.partialInit) {
      ORBTools.kill();
    }
  }

  public
  void
  setName(String name) {
    if (this.pingThread != null) {
      this.disconnect("change of name");
    }
    this.name = name;
  }

  public
  String
  getName() {
    return this.name;
  }

  public
  String
  getHostname() {
    return this.hostname;
  }

  public
  short
  connect(String message) {
    if (this.pingThread != null) {
      this.disconnect("for immediate reconnection");
    }
    short ret;
    StringHolder s = new StringHolder(this.name);
    long epoch = System.currentTimeMillis();
    long sec = epoch / 1000;
    long msec = epoch - (sec * 1000);
    log_time_t lt = new log_time_t(sec, (int)msec);
    String[] initialConfigValue = new String[0];
    tag_list_tHolder initialConfig = new tag_list_tHolder(initialConfigValue);
    try {
      ret = this.LCCref.connectComponent(
        s,
        this.hostname,
        message,
        this.CCref,
        lt,
        initialConfig
      );
    } catch (Exception ex) {
      return 1;
    }
    this.name = s.value;
    this.currentTagList = initialConfig.value;
    this.pingThread = new PingThread(this);
    this.flushBufferThread = new FlushBufferThread(this);
    this.pingThread.startThread();
    this.flushBufferThread.startThread();
    return ret;
  }

  public
  short
  disconnect(String message) {
    if (this.pingThread == null) {
      return LS_OK.value;
    }
    this.pingThread.stopThread();
    this.flushBufferThread.stopThread();
    this.pingThread = null;
    this.flushBufferThread = null;
    short ret;
    try {
      ret = this.LCCref.disconnectComponent(this.name, message);
    } catch (Exception ex) {
      ret = 1;
    }
    return 1;
  }

  public
  void
  sendMsg(String tag, String msg) {
    long epoch = System.currentTimeMillis();
    long sec = epoch / 1000;
    long msec = epoch - (sec * 1000);
    log_time_t lt = new log_time_t(sec, (int)msec);
    log_msg_t logmsg = new log_msg_t(
      this.name,
      lt,
      false,
      tag,
      msg
    );
    log_msg_t[] list = new log_msg_t[this.buffer.length + 1];
    System.arraycopy(this.buffer, 0, list, 0, this.buffer.length);
    list[this.buffer.length] = logmsg;
    this.buffer = list;
  }

  public
  String
  getHostnameOf(String componentName) {
    String ret;
    try {
      ret = this.LCCref.getHostnameOf(componentName);
    } catch (Exception ex) {
      ret = null;
    }
    return ret;
  }

  public
  boolean
  isLog(String tagname) {
    ArrayList list = new ArrayList(Arrays.asList(this.currentTagList));
    return list.contains(tagname);
  }

  public
  LogCentralComponent LCCref;

  public
  String[] currentTagList;

  public
  log_msg_t[] buffer;



  private
  ComponentConfiguratorImpl CCimpl;

  private
  ComponentConfigurator CCref;

  private
  boolean partialInit;

  private
  String name;

  private
  String hostname;

  private
  PingThread pingThread;

  private
  FlushBufferThread flushBufferThread;

  private
  byte[] id;
}

File: src_clean/log/libraries/LogComponentBaseC++/LogComponentBase.cc
/****************************************************************************/
/* LogComponentBase implementation class                                    */
/*                                                                          */
/*  Author(s):                                                              */
/*    - Georg Hoesch (hoesch@in.tum.de)                                     */
/*    - Cyrille Pontvieux (cyrille.pontvieux@edu.univ-fcomte.fr)            */
/*                                                                          */
/* $LICENSE$                                                                */
/****************************************************************************/
/* $Id: LogComponentBase.cc,v 1.8 2011/02/04 15:56:03 bdepardo Exp $
 * $Log: LogComponentBase.cc,v $
 * Revision 1.8  2011/02/04 15:56:03  bdepardo
 * Add missing include
 *
 * Revision 1.7  2011/02/04 15:53:55  bdepardo
 * Use a const char* name in constructor, then strdup the name.
 * Then use free instead of delete on this->name
 *
 * Revision 1.6  2011/02/04 15:21:50  bdepardo
 * delete [] hostname instead of delete hostname
 *
 * Revision 1.5  2010/12/03 12:40:26  kcoulomb
 * MAJ log to use forwarders
 *
 * Revision 1.4  2010/11/10 04:24:50  bdepardo
 * LogServiceC instead of LogService
 *
 * Revision 1.3  2010/11/10 02:27:44  kcoulomb
 * Update the log to use the forwarder.
 * Programm run without launching forwarders but fails with forwarder.
 *
 * Revision 1.2  2007/08/31 16:41:16  bdepardo
 * When trying to add a new component, we check if the name of the component exists and if the component is reachable
 * - it the name already exists:
 *    - if the component is reachable, then we do not connect the new component
 *    - else we consider that the component is lost, and we delete the old component ant add the new one
 * - else add the component
 *
 * Revision 1.1  2004/01/09 11:07:12  ghoesch
 * Restructured the whole LogService source tree.
 * Added autotools make process. Cleaned up code.
 * Removed some testers. Ready to release.
 *
 ****************************************************************************/

#include <cstdlib>
#include <unistd.h>
#include <sys/time.h>
#include <iostream>

#include "LogComponentBase.hh"
#include "ORBTools.hh"
#include "ORBMgr.hh"

using namespace std;

LogComponentBase::LogComponentBase(bool* success, int argc, char** argv,
                                   unsigned int tracelevel, const char* name,
                                   unsigned int port) {
  *success = false;
  this->name = strdup(name);
  this->hostname = new char[256];
  if (gethostname(hostname, 255)) {
    return;
  }
  if (!ORBTools::init(argc, argv, tracelevel, port)) {
    return;
  }
  CORBA::Object* object;
  if (!ORBTools::findServant("LogServiceC", "", "LCC", "", object)) {
    cerr << "Cannot contact the LogComponent servant of LogService!\n";
    return;
  }
  this->LCCref = LogCentralComponent::_narrow(object);
  this->CCimpl = new ComponentConfigurator_impl(this);
  if (!ORBTools::activateServant(this->CCimpl)) {
    cerr << "Cannot activate the ComponentConfigurator servant !\n";
    return;
  }
  if (!ORBTools::activatePOA()) {
    return;
  }
  this->CCref = this->CCimpl->_this();
  this->CCimpl = NULL;
  this->pingThread = NULL;
  this->flushBufferThread = NULL;

  ORBMgr::getMgr()->bind("LogServiceC", getName(), this->CCref, true);
  ORBMgr::getMgr()->fwdsBind("LogServiceC", getName(),
                             ORBMgr::getMgr()->getIOR(this->CCref));
  // Comment because test is empty
  //  try { // To make sure that the object is callable
  //    //    this->LCCref->test();
  //  } catch (...) {
  //    return;
  //  }
  *success = true;
}

LogComponentBase::~LogComponentBase()
{
  if (this->pingThread != NULL) {
    this->pingThread->stopThread();
  }
  if (this->flushBufferThread != NULL) {
    this->flushBufferThread->stopThread();
  }
  free(this->name);
  delete [] this->hostname;
  ORBTools::kill();
}

void
LogComponentBase::setName(const char* name)
{
  if (this->pingThread != NULL) {
    this->disconnect("change of name");
  }
  free(this->name);
  this->name = strdup(name);
}

char*
LogComponentBase::getName()
{
  return strdup(this->name);
}

char*
LogComponentBase::getHostname()
{
  return strdup(this->hostname);
}

short
LogComponentBase::connect(const char* message)
{
  if (this->pingThread != NULL) {
    this->disconnect("for immediate reconnection");
  }
  short ret;
  CORBA::String_var corbaname = CORBA::string_dup(this->name);
  struct timeval tv;
  struct timezone tz;
  gettimeofday(&tv, &tz);
  log_time_t lt;
  lt.sec = tv.tv_sec;
  lt.msec = tv.tv_usec / 1000;
  try {
    ret = this->LCCref->connectComponent(
      corbaname,
      this->hostname,
      message,
      this->name, // MODIF, avant 'this->CC'
      lt,
      this->currentTagList
    );
  } catch (...) {
    return 1;
  }
  //  delete this->name;
  this->name = CORBA::string_dup(corbaname);
  this->pingThread = new PingThread(this);
  this->flushBufferThread = new FlushBufferThread(this);
  this->pingThread->startThread();
  this->flushBufferThread->startThread();
  return ret;
}

short
LogComponentBase::disconnect(const char* message)
{
  if (this->pingThread == NULL) {
    return LS_OK;
  }
  this->pingThread->stopThread();
  this->flushBufferThread->stopThread();
  this->pingThread = NULL;
  this->flushBufferThread = NULL;
  short ret;
  try {
    ret = this->LCCref->disconnectComponent(this->name, message);
  } catch (...) {
    ret = 1;
  }
  return ret;
}

void
LogComponentBase::sendMsg(const char* tag, const char* msg)
{
  log_msg_t logmsg;
  logmsg.componentName = CORBA::string_dup(this->name);
  logmsg.time = this->getLocalTime();
  logmsg.warning = false;
  logmsg.tag = CORBA::string_dup(tag);
  logmsg.msg = CORBA::string_dup(msg);
  CORBA::ULong l = this->buffer.length();
  this->buffer.length(l + 1);
  this->buffer[l] = logmsg;
}

bool
LogComponentBase::isLog(const char* tagname)
{
  bool ret = false;
  CORBA::ULong l = this->currentTagList.length();
  for (CORBA::ULong i = 0 ; i < l ; i++) {
    if (strcmp(tagname, this->currentTagList[i]) == 0) {
      ret = true;
      break;
    }
    if (strcmp("*", this->currentTagList[i]) == 0) {
      ret = true;
      break;
    }
  }
  return ret;
}

log_time_t
LogComponentBase::getLocalTime() {
  struct timeval tv;
  struct timezone tz;
  gettimeofday(&tv, &tz);
  log_time_t ret;
  ret.sec = tv.tv_sec;
  ret.msec = tv.tv_usec / 1000;
  return ret;
}

